File: 02-OCP/cpp/refactored.cpp
#include <iostream>
#include <string>

class Notification {
public:
    virtual void send(const std::string& message) = 0;

    virtual ~Notification() = default;
};


class EmailNotification : public Notification {
public:
    void send(const std::string& message) override {
        std::cout << "Sending Email: "
                  << message << "\n";
    }
};


class SMSNotification : public Notification {
public:
    void send(const std::string& message) override {
        std::cout << "Sending SMS: "
                  << message << "\n";
    }
};


class SlackNotification : public Notification {
public:
    void send(const std::string& message) override {
        std::cout << "Sending Slack message: "
                  << message << "\n";
    }
};


class NotificationService {
public:
    void send(
        Notification& notification,
        const std::string& message
    ) {
        notification.send(message);
    }
};


int main() {
    EmailNotification email;
    SMSNotification sms;
    SlackNotification slack;

    NotificationService service;

    service.send(email, "Order confirmed");
    service.send(sms, "Your OTP is 1234");
    service.send(slack, "Deployment completed");

    return 0;
}

File: 02-OCP/python/refactored.py
from abc import ABC, abstractmethod


class Notification(ABC):

    @abstractmethod
    def send(self, message):
        pass


class EmailNotification(Notification):

    def send(self, message):
        print(f"Sending Email: {message}")


class SMSNotification(Notification):

    def send(self, message):
        print(f"Sending SMS: {message}")


class SlackNotification(Notification):

    def send(self, message):
        print(f"Sending Slack message: {message}")


class NotificationService:

    def send(self, notification, message):
        notification.send(message)


if __name__ == "__main__":
    service = NotificationService()

    service.send(
        EmailNotification(),
        "Order confirmed"
    )

    service.send(
        SMSNotification(),
        "Your OTP is 1234"
    )

    service.send(
        SlackNotification(),
        "Deployment completed"
    )


# Now adding WhatsApp means adding:

# class WhatsAppNotification(Notification):

#     def send(self, message):
#         print(f"Sending WhatsApp: {message}")

# No change to NotificationService. 